Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm

```diff
@@ -0,0 +1,129 @@
+
+NetworkStatus ToNetworkStatus(SCNetworkReachabilityFlags flags) {
+  if (!(flags & kSCNetworkReachabilityFlagsReachable)) {
+    return NetworkStatus::Unreachable;
+  }
+  if (flags & kSCNetworkReachabilityFlagsConnectionRequired) {
+    return NetworkStatus::Unreachable;
+  }
+
+#if TARGET_OS_IPHONE
+  if (flags & kSCNetworkReachabilityFlagsIsWWAN) {
+    return NetworkStatus::ReachableViaCellular;
+  }
+#endif
+  return NetworkStatus::Reachable;
+}
+
+SCNetworkReachabilityRef CreateReachability() {
+  // Pseudoaddress that monitors internet reachability in general.
+  sockaddr_in any_connection_addr{};
+  any_connection_addr.sin_len = sizeof(any_connection_addr);
+  any_connection_addr.sin_family = AF_INET;
+  return SCNetworkReachabilityCreateWithAddress(
+      nullptr, reinterpret_cast<sockaddr*>(&any_connection_addr));
+}
+
+}  // namespace
+
+/**
+ * Implementation of `ConnectivityMonitor` based on `SCNetworkReachability`
+ * (iOS/MacOS).
+ */
+class ConnectivityMonitorApple : public ConnectivityMonitor {
+ public:
+  ConnectivityMonitorApple(AsyncQueue* worker_queue)
+      : ConnectivityMonitor{worker_queue}, reachability_{CreateReachability()} {
+    SCNetworkReachabilityFlags flags;
+    if (SCNetworkReachabilityGetFlags(reachability_, &flags)) {
+      SetInitialStatus(ToNetworkStatus(flags));
+    }
+
+    auto on_reachability_changed = [](SCNetworkReachabilityRef /*unused*/,
+                                      SCNetworkReachabilityFlags flags,
+                                      void* raw_this) {
+      HARD_ASSERT(raw_this, "Received a null pointer as context");
+      static_cast<ConnectivityMonitorApple*>(raw_this)->OnChange(flags);
+    };
+
+    SCNetworkReachabilityContext context{};
+    context.info = this;
+    bool success = SCNetworkReachabilitySetCallback(
+        reachability_, on_reachability_changed, &context);
+    if (!success) {
+      LOG_DEBUG("Couldn't set reachability callback");
+      return;
+    }
+
+    auto queue =
+        dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0);
+    success = SCNetworkReachabilitySetDispatchQueue(reachability_, queue);
+    if (!success) {
+      LOG_DEBUG("Couldn't set reachability queue");
+      return;
+    }
+  }
+
+  ~ConnectivityMonitorApple() {
+    bool success =
+        SCNetworkReachabilitySetCallback(reachability_, nullptr, nullptr);
+    if (!success) {
+      LOG_DEBUG("Couldn't unset reachability callback");
+    }
+    success = SCNetworkReachabilitySetDispatchQueue(reachability_, nullptr);
+    if (!success) {
+      LOG_DEBUG("Couldn't unset reachability queue");
+    }
+  }
+
+ private:
+  void OnChange(SCNetworkReachabilityFlags flags) {
+    MaybeInvokeCallbacks(ToNetworkStatus(flags));
+  }
+
+  SCNetworkReachabilityRef reachability_;
+};
+
+std::unique_ptr<ConnectivityMonitor> ConnectivityMonitor::Create(
+    AsyncQueue* worker_queue) {
+  return absl::make_unique<ConnectivityMonitorApple>(worker_queue);
```

Comment:
I originally thought that each platform would have one implementation file hooked into the build, but now that I think about it, I'm not sure it can work with XCode (I presumed something like `cc_select` in CMake). Happy with any suggestions.

Reply:
If all other means fails, we can have macro guard to each platform implementation and make all into the build (without select at the build level) e.g. foo_android.cc: #if defined(__ANDROID__) #endif  // defined(__ANDROID__), foo_apple.mm: #if defined(__OBJC__) #endif  // defined(__OBJC__) foo_desktop.cc: #if !defined(__ANDROID__) && !defined(__OBJC__) etc.

Reply:
The way this works in Xcode is that we let the podspec match the _apple implementation and [exclude the others in the podspec](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseFirestore.podspec#L44).

Your original strategy of picking an implementation at build time is the right one to follow. You should add `#if defined(__APPLE__)` around this file under the header import.

Reply:
Done. Also added it to the CMake build, and added a `connectivity_monitor_noop.cc` file to make sure the build doesn't break on Linux and Windows.